python/cog/server/http.py

```diff
@@ -88,16 +89,46 @@ def create_app(
 
     predictor_ref = get_predictor_ref(config, mode)
 
+    try:
+        # TODO: avoid loading predictor code in this process
+        predictor = load_predictor_from_ref(predictor_ref)
+        InputType = get_input_type(predictor)
+        OutputType = get_output_type(predictor)
+    except Exception as ex:
+        app.state.health = Health.SETUP_FAILED
+        response = schema.PredictionResponse(input={},
+                                             output=None,
+                                             id=predictor_ref,
+                                             version=None,
+                                             created_at=datetime.now(),
+                                             started_at=datetime.now(),
+                                             completed_at=None,
+                                             logs="",
+                                             error=repr(ex),
+                                             status=schema.Status.FAILED,
+                                             metrics=None)
+        app.state.setup_result_payload = response
+
+        @app.get("/health-check")
+        async def healthcheck_startup_failed() -> Any:
+            return jsonable_encoder(
+                {
+                    "status": app.state.health.name,
+                    "setup": app.state.setup_result_payload,
+                }
+            )
+
+        @app.on_event("shutdown")
+        def shutdown_startup_failed() -> None:
```

Comment:
Hmm, not quite what I meant. I'll leave a longer comment in the PR explaining this.